src/coreclr/src/vm/diagnosticserver.cpp

```diff
@@ -238,4 +255,117 @@ bool DiagnosticServer::Shutdown()
     return fSuccess;
 }
 
+// This method will block runtime bring-up IFF DOTNET_DiagnosticsMonitorAddress != nullptr and DOTNET_DiagnosticsMonitorPauseOnStart!=0 (it's default state)
+// The s_ResumeRuntimeStartupEvent event will be signaled when the Diagnostics Monitor uses the ResumeRuntime Diagnostics IPC Command
+void DiagnosticServer::PauseForDiagnosticsMonitor()
+{
+    CONTRACTL
+    {
+      THROWS;
+      GC_NOTRIGGER;
+      MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
+
+    CLRConfigStringHolder pDotnetDiagnosticsMonitorAddress = CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_DOTNET_DiagnosticsMonitorAddress);
+    if (pDotnetDiagnosticsMonitorAddress != nullptr)
+    {
+        DWORD dwDotnetDiagnosticsMonitorPauseOnStart = CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_DOTNET_DiagnosticsMonitorPauseOnStart);
+        if (dwDotnetDiagnosticsMonitorPauseOnStart != 0)
+        {
+            _ASSERTE(s_ResumeRuntimeStartupEvent != nullptr && s_ResumeRuntimeStartupEvent->IsValid());
+            const DWORD dwFiveSecondWait = s_ResumeRuntimeStartupEvent->Wait(5000, false);
+            if (dwFiveSecondWait == WAIT_TIMEOUT)
+            {
+                STRESS_LOG0(LF_DIAGNOSTICS_PORT, LL_ALWAYS, "The runtime has been configured to pause during startup and is awaiting a Diagnostics IPC ResumeStartup command.");
```

Comment:
I'd still suggest doing this as a printf, unless you got other feedback that we shouldn't or there is a technical challenge?

However if for whatever reason we can't do printf there is no need to wait 5 seconds just for the STRESS_LOG. We can write once before we start waiting and a 2nd message when we stop waiting. If someone sees the first message and not the second it should be pretty apparent where the app stopped making progress.

Reply:
We should be able to use printf here. If you're not seeing the logs immediately, you should be able to use `fflush`